torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -364,26 +330,74 @@ def state_dict(self) -> Dict[str, Any]:
         # - get the pytorch compliant parameter indexing
         state_dict = super().state_dict()
 
-        # - get an id map which links the parameter id to the index in the reference state
-        global_id_map = {id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
-
         # - go through the per-shard states, which are all indexed locally
         for rank, s in enumerate(self._all_states):
             # -- match the local indexing and the global partition, update the corresponding saved state globally
             for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
-                # Go through the parameters indexed locally, pick up the global corresponding param
                 local_index_to_param_id = {
                     i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
                 }
 
                 for local_param_index in local_pg["params"]:
                     # Update the state, if any
                     if local_param_index in s["state"].keys():
-                        global_id = global_id_map[local_index_to_param_id[local_param_index]]
+                        global_id = self.param_to_index[local_index_to_param_id[local_param_index]]
                         state_dict["state"][global_id] = s["state"][local_param_index]
 
         return state_dict
 
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
```

Comment:
minor, but moving that down in the code so that the first functions visible from the top are the one expected from an optimizer + properties specific to this one